[1.14.4/1.15.2] Userdev Login provides incorrect userProperties launch argument

Originally reported [here](https://www.minecraftforge.net/forum/topic/76829-1144-runclient-crash-with-username/) for 1.14.4, another user reported it for 1.15.2 a few hours ago. The problem code appears to be in both versions.

Running the client in userdev with the `--username` and `--password` argument produces [this error](https://scans.gradle.com/s/uiopxxndtcji2/console-log#L764):
```
 java.lang.RuntimeException: java.lang.reflect.InvocationTargetException	
 	at cpw.mods.modlauncher.LaunchServiceHandlerDecorator.launch(LaunchServiceHandlerDecorator.java:39)	
 	at cpw.mods.modlauncher.LaunchServiceHandler.launch(LaunchServiceHandler.java:50)
 	at cpw.mods.modlauncher.LaunchServiceHandler.launch(LaunchServiceHandler.java:68)	
 	at cpw.mods.modlauncher.Launcher.run(Launcher.java:80)	
 	at cpw.mods.modlauncher.Launcher.main(Launcher.java:65)	
 	at net.minecraftforge.userdev.LaunchTesting.main(LaunchTesting.java:101)	
 Caused by: java.lang.reflect.InvocationTargetException	
 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)	
 	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)	
 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)	
 	at java.lang.reflect.Method.invoke(Unknown Source)	
 	at net.minecraftforge.userdev.FMLUserdevClientLaunchProvider.lambda$launchService$0(FMLUserdevClientLaunchProvider.java:55)	
 	at cpw.mods.modlauncher.LaunchServiceHandlerDecorator.launch(LaunchServiceHandlerDecorator.java:37)	
 	... 5 more	
 Caused by: com.google.gson.JsonSyntaxException: com.google.gson.stream.MalformedJsonException: Use JsonReader.setLenient(true) to accept malformed JSON at line 1 column 3 path $.	
 	at com.google.gson.internal.Streams.parse(Streams.java:60)	
 	at com.google.gson.internal.bind.TreeTypeAdapter.read(TreeTypeAdapter.java:65)	
 	at net.minecraft.util.JSONUtils.fromJson(JSONUtils.java:357)	
 	at net.minecraft.util.JSONUtils.fromJson(JSONUtils.java:382)	
 	at net.minecraft.util.JSONUtils.fromJson(JSONUtils.java:403)	
 	at net.minecraft.client.main.Main.main(Main.java:101)	
 	... 11 more	
 Caused by: com.google.gson.stream.MalformedJsonException: Use JsonReader.setLenient(true) to accept malformed JSON at line 1 column 3 path $.	
 	at com.google.gson.stream.JsonReader.syntaxError(JsonReader.java:1559)	
 	at com.google.gson.stream.JsonReader.checkLenient(JsonReader.java:1401)	
 	at com.google.gson.stream.JsonReader.doPeek(JsonReader.java:508)	
 	at com.google.gson.stream.JsonReader.hasNext(JsonReader.java:414)	
 	at com.google.gson.internal.bind.TypeAdapters$29.read(TypeAdapters.java:738)	
 	at com.google.gson.internal.bind.TypeAdapters$29.read(TypeAdapters.java:714)	
 	at com.google.gson.internal.Streams.parse(Streams.java:48)	
 	... 16 more
```

It looks like this is an error in Forge's dev-time login implementation, specifically [this line](https://github.com/MinecraftForge/MinecraftForge/blob/1.15.x/src/userdev/java/net/minecraftforge/userdev/LaunchTesting.java#L163):
```
args.put("userProperties", auth.getUserProperties().toString());
```

As can be seen on [line 12](https://scans.gradle.com/s/uiopxxndtcji2/console-log#L12) of the OP's Gradle scan, this creates a string like `{preferredLanguage=[com.mojang.authlib.properties.Property@d5b810e]}` for the `userProperties` argument.

Minecraft's `Main` class attempts to deserialise the `userProperties` argument as JSON and throws the above error because this string isn't valid JSON.

Instead of using `toString`, `LaunchTesting` probably needs to use `PropertyMap.Serializer` (via `Gson`) to properly serialise the user properties as JSON.

--- a/src/userdev/java/net/minecraftforge/userdev/LaunchTesting.java
+++ b/src/userdev/java/net/minecraftforge/userdev/LaunchTesting.java
@@ -20,9 +20,12 @@
 package net.minecraftforge.userdev;
 
 import com.google.common.base.Strings;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
 import com.mojang.authlib.Agent;
 import com.mojang.authlib.UserAuthentication;
 import com.mojang.authlib.exceptions.AuthenticationException;
+import com.mojang.authlib.properties.PropertyMap;
 import com.mojang.authlib.yggdrasil.YggdrasilAuthenticationService;
 import cpw.mods.modlauncher.Launcher;
 
@@ -157,10 +160,11 @@ public class LaunchTesting
             throw new RuntimeException(e); // don't set other variables
         }
 
+        Gson gson = (new GsonBuilder()).registerTypeAdapter(PropertyMap.class, new PropertyMap.Serializer()).create();
         args.put("username",       auth.getSelectedProfile().getName());
         args.put("uuid",           auth.getSelectedProfile().getId().toString().replace("-", ""));
         args.put("accessToken",    auth.getAuthenticatedToken());
-        args.put("userProperties", auth.getUserProperties().toString());
+        args.put("userProperties", gson.toJson(auth.getUserProperties()));
         return true;
     }
 }

